client/remoterepo.c

```diff
@@ -106,11 +113,490 @@ _handle_curl_cb(
+    const char *digest,
+    const char *type
+    )
+{
+    uint32_t dwError = 0;
+    unsigned char *result;
+    metalink_digest_alias *digest_alias;
+
+    if(IsNullOrEmptyString(filename) ||
+       IsNullOrEmptyString(digest)   ||
+       IsNullOrEmptyString(type))
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+    for(digest_alias = digest_aliases; digest_alias->alias_name;
+        ++digest_alias)
+    {
+        if(!strcmp(digest_alias->alias_name, type))
+        {
+            digest_alias->func(filename, &result);
+            if(!memcmp(result, digest,
+                       digest_alias->length))
+            {
+                /* sha*sum style verdict output */
+                fprintf(stdout, "Validating metalink (%s) OK\n", filename);
+            }
+            else
+            {
+                fprintf(stderr, "Validating metalink (%s) FAILED (digest mismatch)\n",
+                        filename);
+                dwError = ERROR_TDNF_INVALID_REPO_FILE;
+                BAIL_ON_TDNF_SYSTEM_ERROR(dwError); 
+            }
+            break;
+        }
+    }
+
+cleanup:
+    if(result)
+    {
+        TDNF_SAFE_FREE_MEMORY(result);
+    }
+    return dwError;
+error:
+    goto cleanup;
+}
+
+/* Returns nonzero if hex_digest is properly formatted; that is each
+   letter is in [0-9A-Za-z] and the length of the string equals to the
+   result length of digest * 2. */
+uint32_t
+TDNF_Check_Hex_Digest
+    (const char *hex_digest,
+    int digest_length
+    )
+{
+    int i = 0;
+    if(!hex_digest ||
+       !digest_length)
+    {
+        return 0;
+    }
+    for(i = 0; hex_digest[i]; ++i)
+    {
+        char c = hex_digest[i];
+        if(!(('0' <= c && c <= '9') || ('a' <= c && c <= 'z') ||
+             ('A' <= c && c <= 'Z')))
+        {
+            return 0;
+        }
+    }
+    return digest_length * 2 == i;
+}
+
+char 
+TDNF_HexToUint
+    (const char *s)
+{
+    char buf[3];
+    unsigned long val;
+
+    buf[0] = s[0];
+    buf[1] = s[1];
+    buf[2] = 0;
+    val = strtoul(buf, NULL, 16);
+    return (unsigned char)(val&0xff);
+}
+
+uint32_t
+TDNF_New_Metalink_Checksum_From_Hex_Digest
+    (const char *hex_digest,
+    char **ppdigest
+    )
+{
+    uint32_t dwError = 0;
+    char *digest;
+    size_t i;
+    size_t len = 0;
+
+    if(!hex_digest ||
+       !ppdigest)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    len = strlen(hex_digest);
+
+    dwError = TDNFAllocateMemory(1, len/2, (void **)&digest);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    for(i = 0; i < len; i += 2)
+    {
+        digest[i/2] = TDNF_HexToUint(hex_digest + i);
+    }
+    *ppdigest = digest;
+
+cleanup:
+    return dwError;
+
+error:
+    if(digest)
+    {
+        TDNF_SAFE_FREE_MEMORY(digest);
+    }
+    goto cleanup;
+}
+
+uint32_t
+TDNF_New_Metalinkfile
+    (metalink_file_t *fileinfo,
+    metalinkfile ****ml_file
```

Comment:
This '****' construct is strong indicator for a design change/refactor. One possible, first glance fix would be to keep a `metalinkfile *` in `metalink_file_t` but you'd know better.

Generally going above a double pointers is accepted only upto triple pointers :).

Reply:
I am doing the changes to keep it upto double pointer.